File: C1.UWP.Maps/CS/MapsSamples/Samples/VectorLayer/KmlHelper.cs
using C1.C1Zip;
using C1.Xaml.Maps;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Windows.Foundation;

namespace MapsSamples
{
    public delegate bool ProcessVectorItem(C1VectorItemBase vector);

    public static class Extension
    {
        public static void LoadKML(this C1VectorLayer vl, Stream stream, bool centerAndZoom,
          ProcessVectorItem processVector)
        {
            using (StreamReader sr = new StreamReader(stream))
            {
                string s = sr.ReadToEnd();
                List<C1VectorItemBase> vects = KmlReader.Read(s);

                vl.BeginUpdate();
                foreach (C1VectorItemBase vect in vects)
                {
                    if (processVector != null)
                    {
                        if (!processVector(vect))
                            continue;
                    }

                    vl.Children.Add(vect);
                }
                vl.EndUpdate();

                if (centerAndZoom)
                {
                    Rect bnds = vects.GetBounds();
                    C1Maps maps = ((IMapLayer)vl).ParentMaps;

                    if (maps != null)
                    {
                        maps.Center = new Point(bnds.Left + 0.5 * bnds.Width, bnds.Top + 0.5 * bnds.Height);
                        double scale = Math.Max(bnds.Width / 360 * maps.ActualWidth,
                              bnds.Height / 180 * maps.ActualHeight); ;
                        double zoom = Math.Log(512 / scale, 2.0);
                        maps.TargetZoom = maps.Zoom = zoom > 0 ? zoom : 0;
                    }
                }

                sr.Dispose();
            }
        }

        public static void LoadKMZ(this C1VectorLayer vl, Stream stream, bool centerAndZoom,
          ProcessVectorItem processVector)
        {
            using (C1ZipFile zip = new C1ZipFile(stream))
            {
                foreach (C1ZipEntry entry in zip.Entries)
                {
                    if (entry.FileName.EndsWith(".kml"))
                    {
                        using (Stream docStream = entry.OpenReader())
                        {
                            if (docStream != null)
                                LoadKML(vl, docStream, centerAndZoom, processVector);
                        }
                    }
                }
            }
        }

        public static Rect GetBounds(this IEnumerable<C1VectorItemBase> items)
        {
            Rect bounds = Rect.Empty;
            foreach (C1VectorItemBase item in items)
                bounds.Union(item.Bounds);
            return bounds;
        }
    }
}

File: C1.UWP.Maps/VB/MapsSamples/Samples/VectorLayer/KmlHelper.vb
Imports Windows.Foundation
Imports System.Threading.Tasks
Imports System.Text
Imports System.Linq
Imports System.IO
Imports System.Collections.Generic
Imports System
Imports C1.Xaml.Maps
Imports C1.C1Zip

Public Delegate Function ProcessVectorItem(vector As C1VectorItemBase) As Boolean

Public Module Extension
    <Extension()>
    Public Sub LoadKML(vl As C1VectorLayer, stream As Stream, centerAndZoom As Boolean, processVector As ProcessVectorItem)
        Using sr As New StreamReader(stream)
            Dim s As String = sr.ReadToEnd()
            Dim vects As List(Of C1VectorItemBase) = KmlReader.Read(s)

            vl.BeginUpdate()
            For Each vect As C1VectorItemBase In vects
                If processVector IsNot Nothing Then
                    If Not processVector(vect) Then
                        Continue For
                    End If
                End If

                vl.Children.Add(vect)
            Next
            vl.EndUpdate()

            If centerAndZoom Then
                Dim bnds As Rect = vects.GetBounds()
                Dim maps As C1Maps = (CType(vl, IMapLayer)).ParentMaps

                If maps IsNot Nothing Then
                    maps.Center = New Point(bnds.Left + 0.5 * bnds.Width, bnds.Top + 0.5 * bnds.Height)
                    Dim scale As Double = Math.Max(bnds.Width / 360 * maps.ActualWidth, bnds.Height / 180 * maps.ActualHeight)


                    Dim zoom As Double = Math.Log(512 / scale, 2.0)
                    maps.Zoom = If(zoom > 0, zoom, 0)
                    maps.TargetZoom = maps.Zoom
                End If
            End If

            sr.Dispose()
        End Using
    End Sub

    <Extension()>
    Public Sub LoadKMZ(vl As C1VectorLayer, stream As Stream, centerAndZoom As Boolean, processVector As ProcessVectorItem)
        Using zip As New C1ZipFile(stream)
            For Each entry As C1ZipEntry In zip.Entries
                If entry.FileName.EndsWith(".kml") Then
                    Using docStream As Stream = entry.OpenReader()
                        If docStream IsNot Nothing Then
                            LoadKML(vl, docStream, centerAndZoom, processVector)
                        End If
                    End Using
                End If
            Next
        End Using
    End Sub

    <Extension()>
    Public Function GetBounds(items As IEnumerable(Of C1VectorItemBase)) As Rect
        Dim bounds As Rect = Rect.Empty
        For Each item As C1VectorItemBase In items
            bounds.Union(item.Bounds)
        Next
        Return bounds
    End Function
End Module